x/ecocredit/core/genesis.go

```diff
@@ -0,0 +1,267 @@
+				return err
+			}
+
+			if class.Id == classIds[batch.ProjectId] {
+				decimalPlaces[batch.Id] = abbrevToPrecision[class.CreditType]
+			}
+		}
+
+		cItr.Close()
+	}
+
+	calSupplies := make(map[uint64]math.Dec) // map of batchID to calculated supply
+	supplies := make(map[uint64]math.Dec)    // map of batchID to actual supply
+	bsItr, err := ss.BatchSupplyTable().List(ormCtx, api.BatchSupplyPrimaryKey{})
+	if err != nil {
+		return err
+	}
+	defer bsItr.Close()
+
+	for bsItr.Next() {
+		batchSupply, err := bsItr.Value()
+		if err != nil {
+			return err
+		}
+
+		tSupply := math.NewDecFromInt64(0)
+		rSupply := math.NewDecFromInt64(0)
+		if batchSupply.TradableAmount != "" {
+			tSupply, err = math.NewNonNegativeFixedDecFromString(batchSupply.TradableAmount, decimalPlaces[batchSupply.BatchId])
+			if err != nil {
+				return err
+			}
+		}
+		if batchSupply.RetiredAmount != "" {
+			rSupply, err = math.NewNonNegativeFixedDecFromString(batchSupply.RetiredAmount, decimalPlaces[batchSupply.BatchId])
+			if err != nil {
+				return err
+			}
+		}
+
+		total, err := math.SafeAddBalance(tSupply, rSupply)
+		if err != nil {
+			return err
+		}
+
+		supplies[batchSupply.BatchId] = total
+	}
+
+	// calculate credit batch supply from genesis tradable and retired balances
+	if err := calculateSupply(ormCtx, decimalPlaces, ss, calSupplies); err != nil {
+		return err
+	}
+
+	if err := validateSupply(calSupplies, supplies); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func calculateSupply(ctx context.Context, decimalPlaces map[uint64]uint32, ss api.StateStore, calSupply map[uint64]math.Dec) error {
+	bbItr, err := ss.BatchBalanceTable().List(ctx, api.BatchBalancePrimaryKey{})
+	if err != nil {
+		return err
+	}
+	defer bbItr.Close()
+
+	for bbItr.Next() {
+		tBalance := math.NewDecFromInt64(0)
+		rBalance := math.NewDecFromInt64(0)
+
+		balance, err := bbItr.Value()
+		if err != nil {
+			return err
+		}
+
+		if _, ok := decimalPlaces[balance.BatchId]; !ok {
+			return sdkerrors.ErrInvalidType.Wrapf("credit type not exist for %d batch", balance.BatchId)
+		}
+
+		if balance.Tradable != "" {
+			tBalance, err = math.NewNonNegativeFixedDecFromString(balance.Tradable, decimalPlaces[balance.BatchId])
+			if err != nil {
+				return err
+			}
+		}
+
+		if balance.Retired != "" {
+			rBalance, err = math.NewNonNegativeFixedDecFromString(balance.Retired, decimalPlaces[balance.BatchId])
+			if err != nil {
+				return err
+			}
+		}
+
+		total, err := math.Add(tBalance, rBalance)
+		if err != nil {
+			return err
+		}
+
+		if supply, ok := calSupply[balance.BatchId]; ok {
+			result, err := math.SafeAddBalance(supply, total)
+			if err != nil {
+				return err
+			}
+			calSupply[balance.BatchId] = result
+		} else {
+			calSupply[balance.BatchId] = total
+		}
+	}
+
+	return nil
+}
+
+func validateSupply(calSupply, supply map[uint64]math.Dec) error {
+	for denom, cs := range calSupply {
+		if s, ok := supply[denom]; ok {
+			if s.Cmp(cs) != 0 {
```

Comment:
```suggestion
			if s.Cmp(cs) != math.EqualTo {
```

can use this const to make it a lil more readable